## Model Training
#### 1.1 Import Data and Required Packages
Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [2]:
# Basic Import
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
# Modelling
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,AdaBoostRegressor
from sklearn.svm import SVR
from sklearn.linear_model import LinearRegression, Ridge,Lasso
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from sklearn.model_selection import RandomizedSearchCV
# from catboost import CatBoostRegressor
from xgboost import XGBRegressor
import warnings

Import the CSV Data as Pandas DataFrame

In [3]:
df = pd.read_csv('../data/processed/processed_train_set.csv')
test_df = pd.read_csv('../data/processed/processed_test_set.csv')

In [4]:
df.head(5)

inm_floor  inm_size  inm_price  inm_longitude  inm_latitude  his_price  \
0   0.065574  0.035535   0.034237       0.550904      0.713824   0.545495   
1   0.032787  0.023829   0.010545       0.732688      0.303448   0.055181   
2   0.245902  0.081522   0.094034       0.510904      0.700679   0.545495   
3   0.049180  0.054348   0.027009       0.863763      0.793266   0.216906   
4   0.081967  0.035535   0.055659       0.628597      0.715885   0.409451   

   his_quarterly_variation  his_annual_variation  his_monthly_variation  \
0                 0.653710              0.378472               0.360902   
1                 0.522968              0.173611               0.000000   
2                 0.653710              0.378472               0.360902   
3                 0.544170              0.361111               0.157895   
4                 0.650177              0.472222               0.526316   

   dem_Indice_de_reemplazo_de_la_poblacion_activa  ...  \
0                                        0.175559  ...   
1                                        0.411872  ...   
2                                        0.175559  ...   
3                                        0.439892  ...   
4                                        0.392145  ...   

   inm_distrito_Salamanca  inm_distrito_San Blas - Canillejas  \
0                     0.0                                 0.0   
1                     0.0                                 0.0   
2                     0.0                                 0.0   
3                     0.0                                 0.0   
4                     0.0                                 0.0   

   inm_distrito_Tetuán  inm_distrito_Unknown  inm_distrito_Usera  \
0                  0.0                   1.0                 0.0   
1                  0.0                   0.0                 0.0   
2                  0.0                   0.0                 0.0   
3                  0.0                   0.0                 0.0   
4                  0.0                   0.0                 0.0   

   inm_distrito_Vicálvaro  inm_distrito_Villa de Vallecas  \
0                     0.0                             0.0   
1                     0.0                             1.0   
2                     0.0                             0.0   
3                     0.0                             0.0   
4                     0.0                             0.0   

   inm_distrito_Villaverde  inm_barrio_freq  investment_category  
0                      0.0         0.062720                    3  
1                      0.0         0.071882                    0  
2                      0.0         0.062720                    3  
3                      0.0         0.016209                    1  
4                      0.0         0.038055                    2  

[5 rows x 50 columns]

In [5]:
X_train = df.drop('investment_category', axis=1)
y_train = df['investment_category'].astype('category')
X_test = test_df.drop('investment_category', axis=1)
y_test = test_df['investment_category'].astype('category')

In [6]:
# Check the distribution of unique values in the 'investment_category' column
class_distribution = df['investment_category'].value_counts()
# Print the distribution
print(class_distribution)
# Check the data type of the 'investment_category' column
print(df['investment_category'].dtype)

investment_category
0    3573
2    3571
1    3570
3    3500
Name: count, dtype: int64
int64


In [7]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier
# from catboost import CatBoostClassifier  # Uncomment if CatBoost is installed and desired

def evaluate_model(true, predicted):
    accuracy = accuracy_score(true, predicted)
    precision = precision_score(true, predicted, average='weighted')
    recall = recall_score(true, predicted, average='weighted')
    f1 = f1_score(true, predicted, average='weighted')
    return accuracy, precision, recall, f1

# Dictionary of models for classification
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "K-Neighbors Classifier": KNeighborsClassifier(),
    "Decision Tree Classifier": DecisionTreeClassifier(),
    "Random Forest Classifier": RandomForestClassifier(n_estimators=100, random_state=42),
    "XGBClassifier": XGBClassifier(),
    # "CatBoosting Classifier": CatBoostClassifier(verbose=False), # Uncomment if CatBoost is installed and desired
    "AdaBoost Classifier": AdaBoostClassifier(),
    "SVC": SVC()
}

# Lists to store model names and scores
model_list = []
accuracy_list = []
precision_list = []
recall_list = []
f1_list = []

# Loop through models, fit, predict, and evaluate
for name, model in models.items():
    model.fit(X_train, y_train) # Train model
    
    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    # Evaluate Train and Test dataset
    train_accuracy, train_precision, train_recall, train_f1 = evaluate_model(y_train, y_train_pred)
    test_accuracy, test_precision, test_recall, test_f1 = evaluate_model(y_test, y_test_pred)

    print(name)
    model_list.append(name)
    
    print('Model performance for Training set')
    print("- Accuracy: {:.4f}".format(train_accuracy))
    print("- Precision: {:.4f}".format(train_precision))
    print("- Recall: {:.4f}".format(train_recall))
    print("- F1 Score: {:.4f}".format(train_f1))
    print('----------------------------------')
    
    print('Model performance for Test set')
    print("- Accuracy: {:.4f}".format(test_accuracy))
    print("- Precision: {:.4f}".format(test_precision))
    print("- Recall: {:.4f}".format(test_recall))
    print("- F1 Score: {:.4f}".format(test_f1))
    accuracy_list.append(test_accuracy)
    precision_list.append(test_precision)
    recall_list.append(test_recall)
    f1_list.append(test_f1)
    
    print('='*35)
    print('\n')


Logistic Regression
Model performance for Training set
- Accuracy: 0.8811
- Precision: 0.8822
- Recall: 0.8811
- F1 Score: 0.8815
----------------------------------
Model performance for Test set
- Accuracy: 0.8715
- Precision: 0.8730
- Recall: 0.8715
- F1 Score: 0.8721




/Users/adnanbhanji/Documents/GitHub/Madrid-Property-Analysis/venv/lib/python3.9/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.9 is maybe too old for this OS.
  warnings.warn(


K-Neighbors Classifier
Model performance for Training set
- Accuracy: 0.9331
- Precision: 0.9334
- Recall: 0.9331
- F1 Score: 0.9331
----------------------------------
Model performance for Test set
- Accuracy: 0.8943
- Precision: 0.8948
- Recall: 0.8943
- F1 Score: 0.8943


Decision Tree Classifier
Model performance for Training set
- Accuracy: 1.0000
- Precision: 1.0000
- Recall: 1.0000
- F1 Score: 1.0000
----------------------------------
Model performance for Test set
- Accuracy: 1.0000
- Precision: 1.0000
- Recall: 1.0000
- F1 Score: 1.0000


Random Forest Classifier
Model performance for Training set
- Accuracy: 1.0000
- Precision: 1.0000
- Recall: 1.0000
- F1 Score: 1.0000
----------------------------------
Model performance for Test set
- Accuracy: 0.9994
- Precision: 0.9994
- Recall: 0.9994
- F1 Score: 0.9994


XGBClassifier
Model performance for Training set
- Accuracy: 1.0000
- Precision: 1.0000
- Recall: 1.0000
- F1 Score: 1.0000
----------------------------------
Model per

/Users/adnanbhanji/Documents/GitHub/Madrid-Property-Analysis/venv/lib/python3.9/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
/Users/adnanbhanji/Documents/GitHub/Madrid-Property-Analysis/venv/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Users/adnanbhanji/Documents/GitHub/Madrid-Property-Analysis/venv/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"

AdaBoost Classifier
Model performance for Training set
- Accuracy: 0.7488
- Precision: 0.6232
- Recall: 0.7488
- F1 Score: 0.6651
----------------------------------
Model performance for Test set
- Accuracy: 0.7492
- Precision: 0.6238
- Recall: 0.7492
- F1 Score: 0.6656


SVC
Model performance for Training set
- Accuracy: 0.8645
- Precision: 0.8653
- Recall: 0.8645
- F1 Score: 0.8648
----------------------------------
Model performance for Test set
- Accuracy: 0.8592
- Precision: 0.8607
- Recall: 0.8592
- F1 Score: 0.8596




Based on the performance metrics provided, the K-Nearest Neighbors (KNN) Classifier and Logistic Regression emerge as the most suitable models for the Madrid Property Analysis project. These models exhibit a strong balance between accuracy, precision, recall, and F1 score on both the training and test sets, indicating good generalization without overfitting. Specifically, the KNN model shows superior performance metrics on the test set with an accuracy of 0.8943, precision of 0.8948, recall of 0.8943, and an F1 score of 0.8943, which suggests it's capturing the underlying patterns in the data effectively. Logistic Regression also performs well with an accuracy of 0.8715 and similar precision, recall, and F1 score, marking it as a reliable model with the added advantage of being interpretable. These characteristics make both models strong candidates for deploying in a production environment where understanding the model's decision-making process can be as valuable as the prediction accuracy itself. With these models, we can classify the investment potential and recieve good results as shown by the metrics. 